s2/shapeutil_visit_crossing_edge_pairs_test.go

```diff
@@ -0,0 +1,150 @@
+	edgePairs := EdgePairVector{}
+	VisitCrossingEdgePairs(index, crossingType, func(a, b ShapeEdge, _ bool) bool {
+		edgePairs = append(edgePairs, EdgePair{a.ID, b.ID})
+		return true // Continue visiting.
+	})
+	if len(edgePairs) > 1 {
+		sort.Slice(edgePairs, func(i, j int) bool {
+			return edgePairs[i].A.Cmp(edgePairs[j].A) == -1 || edgePairs[i].B.Cmp(edgePairs[j].B) == -1
+		})
+		slices.Compact(edgePairs)
+	}
+	return edgePairs
+}
+
+// Brute force crossings in one index.
+func getCrossingEdgePairsBruteForce(index *ShapeIndex, crossingType CrossingType) EdgePairVector {
+	var result EdgePairVector
+	minSign := Cross
+	if crossingType == CrossingTypeAll {
+		minSign = MaybeCross
+	}
+
+	for aIter := NewEdgeIterator(index); !aIter.Done(); aIter.Next() {
+		a := aIter.Edge()
+		bIter := EdgeIterator{
+			index:    aIter.index,
+			shapeID:  aIter.shapeID,
+			numEdges: aIter.numEdges,
+			edgeID:   aIter.edgeID,
+		}
+		for bIter.Next(); !bIter.Done(); bIter.Next() {
+			b := bIter.Edge()
+			// missinglink: enum ordering is reversed compared to C++
+			if CrossingSign(a.V0, a.V1, b.V0, b.V1) <= minSign {
+				result = append(result, EdgePair{
+					aIter.ShapeEdgeID(),
+					bIter.ShapeEdgeID(),
+				})
+			}
+		}
+	}
+	return result
+}
+
+func TestGetCrossingEdgePairs(t *testing.T) {
+	var index ShapeIndex
+	if len(getCrossings(&index, CrossingTypeAll)) != 0 {
+		t.Error("Expected 0 crossings in empty index")
+	}
+	if len(getCrossings(&index, CrossingTypeInterior)) != 0 {
+		t.Error("Expected 0 interior crossings in empty index")
+	}
+}
+
+func TestGetCrossingEdgePairsGrid(t *testing.T) {
+	kGridSize := 10.0
+  epsilon := 1e-10
+
+	// There are 11 horizontal and 11 vertical lines. The expected number of
+  // interior crossings is 9x9, plus 9 "touching" intersections along each of
+  // the left, right, and bottom edges. "epsilon" is used to make the interior
+  // lines slightly longer so the "touches" actually cross, otherwise 3 of the
+  // 27 touches are not considered intersecting.
+  // However, the vertical lines do not reach the top line as it curves on the
+  // surface of the sphere: despite "epsilon" those 9 are not even very close
+  // to intersecting. Thus 9 * 12 = 108 interior and four more at the corners
+  // when CrossingType::ALL is used.
+
+	index :=  NewShapeIndex()
+	shape := edgeVectorShape{}
+
+  for i := 0.0; i <= kGridSize; i++ {
+    var e = epsilon
+		if i == 0 || i == kGridSize {
+			e = 0
+		}
+
+    shape.Add(PointFromLatLng(LatLngFromDegrees(-e, i)), PointFromLatLng(LatLngFromDegrees(kGridSize + e, i)));
+    shape.Add(PointFromLatLng(LatLngFromDegrees(i, -e)), PointFromLatLng(LatLngFromDegrees(i, kGridSize + e)));
+  }
+
+	index.Add(&shape)
+	if len(getCrossingEdgePairsBruteForce(index, CrossingTypeAll)) != 112 {
+		t.Errorf("Fail")
+	}
+	if len(getCrossingEdgePairsBruteForce(index, CrossingTypeInterior)) != 108 {
+		t.Errorf("Fail")
+	}
+}
+
```

Comment:
There are none of the two index versions of the tests.  Can you add a TODO for them to be added?